src/hotspot/cpu/riscv/assembler_riscv_c.hpp

```diff
@@ -0,0 +1,581 @@
+  }
+
+  INSN(c_lwsp, 0b010, 0b10);
+
+#undef INSN
+
+#define INSN(NAME, funct3, op)                                                               \
+  void NAME(Register Rd_Rs2, Register Rs1, uint32_t uimm) {                                  \
+    assert_cond(is_unsigned_imm_in_range(uimm, 7, 0));                                       \
+    assert_cond((uimm & 0b11) == 0);                                                         \
+    uint16_t insn = 0;                                                                       \
+    c_patch((address)&insn, 1, 0, op);                                                       \
+    c_patch_compressed_reg((address)&insn, 2, Rd_Rs2);                                       \
+    c_patch((address)&insn, 5, 5, (uimm & nth_bit(6)) >> 6);                                 \
+    c_patch((address)&insn, 6, 6, (uimm & nth_bit(2)) >> 2);                                 \
+    c_patch_compressed_reg((address)&insn, 7, Rs1);                                          \
+    c_patch((address)&insn, 12, 10, (uimm & right_n_bits(6)) >> 3);                          \
+    c_patch((address)&insn, 15, 13, funct3);                                                 \
+    emit_int16(insn);                                                                        \
+  }
+
+  INSN(c_lw, 0b010, 0b00);
+  INSN(c_sw, 0b110, 0b00);
+
+#undef INSN
+
+#define INSN(NAME, funct3, op)                                                               \
+  void NAME() {                                                                              \
+    uint16_t insn = 0;                                                                       \
+    c_patch((address)&insn, 1, 0, op);                                                       \
+    c_patch((address)&insn, 11, 2, 0x0);                                                     \
+    c_patch((address)&insn, 12, 12, 0b1);                                                    \
+    c_patch((address)&insn, 15, 13, funct3);                                                 \
+    emit_int16(insn);                                                                        \
+  }
+
+  INSN(c_ebreak, 0b100, 0b10);
+
+#undef INSN
+
+public:
+// C-Ext: an abstact compressible region
+class AbstractCompressibleRegion : public StackObj {
+protected:
+  Assembler *_masm;
+  bool _prev_in_compressible_region;
+protected:
+  AbstractCompressibleRegion(Assembler *_masm)
+  : _masm(_masm)
+  , _prev_in_compressible_region(_masm->in_compressible_region()) {}
+};
+
+class CompressibleRegion : public AbstractCompressibleRegion {
+public:
+  CompressibleRegion(Assembler *_masm) : AbstractCompressibleRegion(_masm) {
+    _masm->set_in_compressible_region(true);
+  }
+  ~CompressibleRegion() {
+    _masm->set_in_compressible_region(_prev_in_compressible_region);
+  }
+};
+
+// C-Ext: an uncompressible region
+class UncompressibleRegion : public AbstractCompressibleRegion {
```

Comment:
It looks strange that UncompressibleRegion extends the abstract 'compressible' region here.